File: java/com/epicdragonworld/gameserver/network/GameClient.java
package com.epicdragonworld.gameserver.network;

import java.util.logging.Logger;

import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;

import com.epicdragonworld.gameserver.managers.WorldManager;
import com.epicdragonworld.gameserver.model.actor.Player;

/**
 * @author Pantelis Andrianakis
 */
public class GameClient extends SimpleChannelInboundHandler<byte[]>
{
	private static final Logger LOGGER = Logger.getLogger(GameClient.class.getName());
	
	private Channel _channel;
	private String _ip;
	private String _accountName;
	private Player _activeChar;
	
	@Override
	public void handlerAdded(ChannelHandlerContext ctx)
	{
		// Connected.
		_channel = ctx.channel();
		_ip = _channel.remoteAddress().toString();
		_ip = _ip.substring(1, _ip.lastIndexOf(':')); // Trim out /127.0.0.1:12345
	}
	
	public void channelSend(SendablePacket packet)
	{
		if (_channel.isActive())
		{
			_channel.writeAndFlush(packet.getSendableBytes());
		}
	}
	
	@Override
	protected void channelRead0(ChannelHandlerContext ctx, byte[] bytes)
	{
		RecievablePacketManager.handle(this, new ReceivablePacket(Encryption.decrypt(bytes)));
	}
	
	@Override
	public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
	{
	}
	
	@Override
	public void channelInactive(ChannelHandlerContext ctx)
	{
		// Disconnected.
		WorldManager.removeClient(this);
		LOGGER.finer("Client Disconnected: " + ctx.channel());
	}
	
	public String getIp()
	{
		return _ip;
	}
	
	public String getAccountName()
	{
		return _accountName;
	}
	
	public void setAccountName(String accountName)
	{
		_accountName = accountName;
	}
	
	public Player getActiveChar()
	{
		return _activeChar;
	}
	
	public void setActiveChar(Player activeChar)
	{
		_activeChar = activeChar;
	}
}

File: EpicDragonWorldServer/Network/GameClient.cs
using System;
using System.Net.Sockets;

/**
 * Author: Pantelis Andrianakis
 * Date: November 7th 2018
 */
public class GameClient
{
    private readonly NetworkStream networkStream;
    private readonly string ip;
    private string accountName = "";
    private Player activeChar;

    public GameClient(NetworkStream stream, string address)
    {
        networkStream = stream;
        // Clean IP address.
        ip = address.Substring(0, address.LastIndexOf(']'));
        int start = ip.LastIndexOf(':') + 1;
        ip = ip.Substring(start, ip.Length - start);
    }

    public async void ChannelSend(SendablePacket packet)
    {
        try
        {
            await networkStream.WriteAsync(packet.GetSendableBytes());
        }
        catch (Exception)
        {
            // Connection closed from client side.
        }
    }

    public string GetIp()
    {
        return ip;
    }

    public string GetAccountName()
    {
        return accountName;
    }

    public void SetAccountName(string name)
    {
        accountName = name;
    }

    public Player GetActiveChar()
    {
        return activeChar;
    }

    public void SetActiveChar(Player activeChar)
    {
        this.activeChar = activeChar;
    }
}
